File: src/generation/rollback.zig
const std = @import("std");
const fmt = std.fmt;
const fs = std.fs;
const mem = std.mem;
const posix = std.posix;
const linux = std.os.linux;
const Allocator = mem.Allocator;
const ArrayList = std.ArrayList;
const ArgIterator = std.process.ArgIterator;

const argparse = @import("../argparse.zig");
const argError = argparse.argError;
const argIs = argparse.argIs;
const getNextArgs = argparse.getNextArgs;
const ArgParseError = argparse.ArgParseError;

const config = @import("../config.zig");

const Constants = @import("../constants.zig");

const log = @import("../log.zig");

const utils = @import("../utils.zig");
const fileExistsAbsolute = utils.fileExistsAbsolute;
const runCmd = utils.runCmd;

// Reusing specialization logic from `nixos apply`, because I'm lazy.
const findSpecialization = @import("../apply.zig").findSpecialization;

pub const GenerationRollbackCommand = struct {
    verbose: bool = false,
    dry: bool = false,
    specialization: ?[]const u8 = null,
    yes: bool = false,

    const usage =
        \\Rollback to the previous NixOS generation.
        \\
        \\Usage:
        \\    nixos generation rollback [options]
        \\
        \\Options:
        \\    -d, --dry               Show what would be activated, but do not activate
        \\    -h, --help              Show this help menu
        \\    -s, --specialisation    Activate the given specialisation
        \\    -v, --verbose           Show verbose logging
        \\    -y, --yes               Automatically confirm activation
        \\
    ;

    pub fn parseArgs(argv: *ArgIterator, parsed: *GenerationRollbackCommand) !?[]const u8 {
        var next_arg = argv.next();
        while (next_arg) |arg| {
            if (argIs(arg, "--dry", "-d")) {
                parsed.dry = true;
            } else if (argIs(arg, "--help", "-h")) {
                log.print("{s}", .{usage});
                return ArgParseError.HelpInvoked;
            } else if (argIs(arg, "--specialisation", "-s")) {
                const next = (try getNextArgs(argv, arg, 1))[0];
                parsed.specialization = next;
            } else if (argIs(arg, "--verbose", "-v")) {
                parsed.verbose = true;
            } else if (argIs(arg, "--yes", "-y")) {
                parsed.yes = true;
            } else {
                return arg;
            }

            next_arg = argv.next();
        }

        return null;
    }
};

const GenerationRollbackError = error{
    PermissionDenied,
    ResourceAccessFailed,
    SetNixProfileFailed,
    SwitchToConfigurationFailed,
} || Allocator.Error;

var exit_status: u8 = 0;
var verbose: bool = false;

pub fn setNixEnvProfile(allocator: Allocator, profile_dirname: []const u8, dry: bool, gen_number: usize) !void {
    const gen_number_str = try fmt.allocPrint(allocator, "{d}", .{gen_number});
    defer allocator.free(gen_number_str);

    const argv = &.{ "nix-env", "--profile", profile_dirname, "--switch-generation", gen_number_str };

    if (verbose) log.cmd(argv);

    if (dry) return;

    const result = runCmd(.{
        .allocator = allocator,
        .argv = argv,
    }) catch return GenerationRollbackError.SetNixProfileFailed;

    if (result.status != 0) {
        exit_status = result.status;
        return GenerationRollbackError.SetNixProfileFailed;
    }
}

fn runSwitchToConfiguration(
    allocator: Allocator,
    location: []const u8,
    command: []const u8,
) !void {
    const argv = &.{ location, command };

    if (verbose) log.cmd(argv);

    const result = runCmd(.{
        .allocator = allocator,
        .argv = argv,
    }) catch return GenerationRollbackError.SwitchToConfigurationFailed;

    if (result.status != 0) {
        exit_status = result.status;
        return GenerationRollbackError.SwitchToConfigurationFailed;
    }
}

fn rollbackGeneration(allocator: Allocator, args: GenerationRollbackCommand, profile_name: []const u8) !void {
    verbose = args.verbose;

    const c = config.getConfig();

    if (linux.geteuid() != 0) {
        utils.execAsRoot(allocator) catch |err| {
            log.err("unable to re-exec this command as root: {s}", .{@errorName(err)});
            return GenerationRollbackError.PermissionDenied;
        };
    }

    const base_profile_dirname = if (mem.eql(u8, profile_name, "system"))
        Constants.nix_profiles
    else
        Constants.nix_system_profiles;
    const profile_dirname = try fs.path.join(allocator, &.{ base_profile_dirname, profile_name });
    defer allocator.free(profile_dirname);

    // While it is possible to use the `rollback` command, we still need
    // to find the previous generation number ourselves in order to run
    // `nvd` or `nix store diff-closures` properly.
    const gen_list = utils.generation.gatherGenerationsFromProfile(allocator, profile_name) catch return GenerationRollbackError.ResourceAccessFailed;
    defer allocator.free(gen_list);

    const current_gen_idx = blk: {
        for (gen_list, 0..) |gen, i| {
            if (gen.current) break :blk i;
        }

        log.err("no current generation detected in generations list", .{});
        return GenerationRollbackError.ResourceAccessFailed;
    };

    if (current_gen_idx == 0) {
        log.err("no generation older than the current one ({d}) exists", .{gen_list[current_gen_idx].generation.?});
        return GenerationRollbackError.ResourceAccessFailed;
    }

    const prev_gen = gen_list[current_gen_idx - 1];
    const prev_gen_number = prev_gen.generation.?;

    const prev_gen_dirname = try fmt.allocPrint(allocator, "{s}/{s}-{d}-link", .{ base_profile_dirname, profile_name, prev_gen_number });
    defer allocator.free(prev_gen_dirname);

    log.step("Comparing changes...", .{});
    const diff_cmd_status = utils.generation.diff(allocator, Constants.current_system, prev_gen_dirname, verbose) catch |err| blk: {
        log.warn("diff command failed to run: {s}", .{@errorName(err)});
        break :blk 0;
    };
    if (diff_cmd_status != 0) {
        log.warn("diff command exited with status {d}", .{diff_cmd_status});
    }

    // Ask for confirmation, if needed
    if (!args.yes and !c.no_confirm) {
        log.print("\n", .{});
        const confirm = utils.confirmationInput("Activate previous generation") catch |err| {
            log.err("unable to read stdin for confirmation: {s}", .{@errorName(err)});
            return GenerationRollbackError.ResourceAccessFailed;
        };
        if (!confirm) {
            log.warn("confirmation was not given, not proceeding with activation", .{});
            return;
        }
    }

    log.step("Activating previous generation...", .{});

    // Rollback and set generation profile
    setNixEnvProfile(allocator, profile_dirname, args.dry, prev_gen_number) catch |err| {
        log.err("failed to set system profile with nix-env", .{});
        return err;
    };

    // Switch to configuration
    const specialization = args.specialization orelse findSpecialization(allocator) catch blk: {
        log.warn("using base configuration without specialisations", .{});
        break :blk null;
    };

    const stc = if (specialization) |spec|
        try fs.path.join(allocator, &.{ profile_dirname, "specialisation", spec, "/bin/switch-to-configuration" })
    else
        try fs.path.join(allocator, &.{ profile_dirname, "/bin/switch-to-configuration" });
    defer allocator.free(stc);

    if (specialization) |spec| {
        if (!fileExistsAbsolute(stc)) {
            log.warn("could not find specialisation '{s}'", .{spec});
            log.warn("using base configuration without specialisations", .{});
        }
    }

    const action = if (args.dry) "dry-activate" else "switch";
    try runSwitchToConfiguration(allocator, stc, action);
}

pub fn generationRollbackMain(allocator: Allocator, args: GenerationRollbackCommand, profile: ?[]const u8) u8 {
    const profile_name = profile orelse "system";

    rollbackGeneration(allocator, args, profile_name) catch |err| {
        switch (err) {
            GenerationRollbackError.PermissionDenied => return 13,
            GenerationRollbackError.ResourceAccessFailed => return 3,
            GenerationRollbackError.SetNixProfileFailed, GenerationRollbackError.SwitchToConfigurationFailed => {
                return if (exit_status != 0) exit_status else 1;
            },
            Allocator.Error.OutOfMemory => {
                log.err("out of memory, cannot continue", .{});
                return 1;
            },
        }
    };

    return 0;
}

File: cmd/generation/rollback/rollback.go
package rollback

import (
	"fmt"
	"os"
	"path/filepath"
	"slices"

	"github.com/spf13/cobra"

	genUtils "github.com/water-sucks/nixos/cmd/generation/shared"
	"github.com/water-sucks/nixos/internal/activation"
	cmdTypes "github.com/water-sucks/nixos/internal/cmd/types"
	cmdUtils "github.com/water-sucks/nixos/internal/cmd/utils"
	"github.com/water-sucks/nixos/internal/constants"
	"github.com/water-sucks/nixos/internal/generation"
	"github.com/water-sucks/nixos/internal/logger"
	"github.com/water-sucks/nixos/internal/settings"
	"github.com/water-sucks/nixos/internal/system"
	"github.com/water-sucks/nixos/internal/utils"
)

func GenerationRollbackCommand(genOpts *cmdTypes.GenerationOpts) *cobra.Command {
	opts := cmdTypes.GenerationRollbackOpts{}

	cmd := cobra.Command{
		Use:   "rollback [flags] {GEN}",
		Short: "Activate the previous generation",
		Long:  "Rollback to the previous NixOS generation.",
		RunE: func(cmd *cobra.Command, args []string) error {
			return cmdUtils.CommandErrorHandler(generationRollbackMain(cmd, genOpts, &opts))
		},
	}

	cmd.Flags().BoolVarP(&opts.Dry, "dry", "d", false, "Show what would be activated, but do not activate")
	cmd.Flags().StringVarP(&opts.Specialisation, "specialisation", "s", "", "Activate the specialisation with `name`")
	cmd.Flags().BoolVarP(&opts.Verbose, "verbose", "v", false, "Show verbose logging")
	cmd.Flags().BoolVarP(&opts.AlwaysConfirm, "yes", "y", false, "Automatically confirm activation")

	_ = cmd.RegisterFlagCompletionFunc("specialisation", completeSpecialisationFlag(genOpts.ProfileName))

	cmdUtils.SetHelpFlagText(&cmd)

	return &cmd
}

func generationRollbackMain(cmd *cobra.Command, genOpts *cmdTypes.GenerationOpts, opts *cmdTypes.GenerationRollbackOpts) error {
	log := logger.FromContext(cmd.Context())
	cfg := settings.FromContext(cmd.Context())
	s := system.NewLocalSystem(log)

	if os.Geteuid() != 0 {
		err := utils.ExecAsRoot(cfg.RootCommand)
		if err != nil {
			log.Errorf("failed to re-exec command as root: %v", err)
			return err
		}
	}

	// While it is possible to use the `rollback` command, we still need
	// to find the previous generation number ourselves in order to run
	// `nvd` or `nix store diff-closures` properly.
	previousGen, err := findPreviousGeneration(log, genOpts.ProfileName)
	if err != nil {
		return err
	}

	profileDirectory := constants.NixProfileDirectory
	if genOpts.ProfileName != "system" {
		profileDirectory = constants.NixSystemProfileDirectory
	}
	generationLink := filepath.Join(profileDirectory, fmt.Sprintf("%v-%v-link", genOpts.ProfileName, previousGen.Number))

	log.Step("Comparing changes...")

	err = generation.RunDiffCommand(log, s, constants.CurrentSystem, generationLink, &generation.DiffCommandOptions{
		UseNvd:  cfg.UseNvd,
		Verbose: opts.Verbose,
	})
	if err != nil {
		log.Errorf("failed to run diff command: %v", err)
	}

	if !opts.AlwaysConfirm {
		log.Printf("\n")
		confirm, err := cmdUtils.ConfirmationInput("Activate the previous generation?")
		if err != nil {
			log.Errorf("failed to get confirmation: %v", err)
			return err
		}
		if !confirm {
			msg := "confirmation was not given, skipping activation"
			log.Warn(msg)
			return fmt.Errorf("%v", msg)
		}
	}

	specialisation := opts.Specialisation
	if specialisation == "" {
		defaultSpecialisation, err := activation.FindDefaultSpecialisationFromConfig(generationLink)
		if err != nil {
			log.Warnf("unable to find default specialisation from config: %v", err)
		} else {
			specialisation = defaultSpecialisation
		}
	}

	if !activation.VerifySpecialisationExists(generationLink, specialisation) {
		log.Warnf("specialisation '%v' does not exist", specialisation)
		log.Warn("using base configuration without specialisations")
		specialisation = ""
	}

	previousGenNumber, err := activation.GetCurrentGenerationNumber(genOpts.ProfileName)
	if err != nil {
		log.Errorf("%v", err)
		return err
	}

	if !opts.Dry {
		log.Step("Setting system profile...")

		if err := activation.SetNixProfileGeneration(s, genOpts.ProfileName, uint64(previousGen.Number), opts.Verbose); err != nil {
			log.Errorf("failed to set system profile: %v", err)
			return err
		}
	}

	// In case switch-to-configuration fails, rollback the profile.
	// This is to prevent accidental deletion of all working
	// generations in case the switch-to-configuration script
	// fails, since the active profile will not be rolled back
	// automatically.
	rollbackProfile := false
	if !opts.Dry {
		defer func(rollback *bool) {
			if !*rollback {
				return
			}

			if !cfg.AutoRollback {
				log.Warnf("automatic rollback is disabled, the currently active profile may have unresolved problems")
				log.Warnf("you are on your own!")
				return
			}

			log.Step("Rolling back system profile...")
			if err := activation.SetNixProfileGeneration(s, "system", previousGenNumber, opts.Verbose); err != nil {
				log.Errorf("failed to rollback system profile: %v", err)
				log.Info("make sure to rollback the system manually before deleting anything!")
			}
		}(&rollbackProfile)
	}

	log.Step("Activating...")

	var stcAction activation.SwitchToConfigurationAction = activation.SwitchToConfigurationActionSwitch
	if opts.Dry {
		stcAction = activation.SwitchToConfigurationActionDryActivate
	}

	err = activation.SwitchToConfiguration(s, generationLink, stcAction, &activation.SwitchToConfigurationOptions{
		Verbose:        opts.Verbose,
		Specialisation: specialisation,
	})
	if err != nil {
		rollbackProfile = true
		log.Errorf("failed to switch to configuration: %v", err)
		return err
	}

	return nil
}

func findPreviousGeneration(log *logger.Logger, profileName string) (*generation.Generation, error) {
	generations, err := genUtils.LoadGenerations(log, profileName, false)
	if err != nil {
		return nil, err
	}

	currentGenIdx := slices.IndexFunc(generations, func(g generation.Generation) bool {
		return g.IsCurrent
	})
	if currentGenIdx == -1 {
		panic("current generation not found, this is a bug")
	}
	currentGen := generations[currentGenIdx]

	if currentGenIdx == 0 {
		msg := fmt.Sprintf("no generation older than the current one (%v) exists", currentGen.Number)
		log.Error(msg)
		return nil, fmt.Errorf("%v", msg)
	}

	return &generations[currentGenIdx-1], nil
}

func completeSpecialisationFlag(profileName string) cmdTypes.CompletionFunc {
	profileDirectory := constants.NixProfileDirectory
	if profileName != "system" {
		profileDirectory = constants.NixSystemProfileDirectory
	}

	return func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
		// I was too lazy to not
		log := logger.FromContext(cmd.Context())

		previousGen, err := findPreviousGeneration(log, profileName)
		if err != nil {
			return []string{}, cobra.ShellCompDirectiveNoFileComp
		}

		generationLink := filepath.Join(profileDirectory, fmt.Sprintf("%v-%v-link", profileName, previousGen.Number))

		return generation.CompleteSpecialisationFlag(generationLink)(cmd, args, toComplete)
	}
}
